src/transforms/groupby.js

```diff
@@ -158,20 +152,52 @@ function transformOne(trace, state) {
         styleLookup[styles[i].target] = styles[i].value;
     }
 
+    var newDataByGroup = {};
+
     for(i = 0; i < groupNames.length; i++) {
-        var groupName = groupNames[i];
+        groupName = groupNames[i];
+
+        // Start with a deep extend that just copies array references.
+        newTrace = newData[i] = newDataByGroup[groupName] = Lib.extendDeepNoArrays({}, trace);
+        newTrace.name = groupName;
 
-        var newTrace = newData[i] = Lib.extendDeepNoArrays({}, trace);
+        // In order for groups to apply correctly to other transform data (e.g.
+        // a filter transform), we have to break the connection and clone the
+        // transforms so that each group writes grouped values into a different
+        // destination. This function does not break the array reference
+        // connection between the split transforms it creates. That's handled in
+        // initialize, which creates a new empty array for each arrayAttr.
+        transforms = newTrace.transforms;
+        newTrace.transforms = [];
+        for(j = 0; j < transforms.length; j++) {
+            newTrace.transforms[j] = Lib.extendDeepNoArrays({}, transforms[j]);
+        }
+
+        // Initialize empty arrays for the arrayAttrs, to be split in the next step
+        for(j = 0; j < arrayAttrs.length; j++) {
+            Lib.nestedProperty(newTrace, arrayAttrs[j]).set([]);
```

Comment:
nice - what you're doing here is pretty similar in fact to what I'm doing with the aggregate transform 🎉 

I suspect you could speed it up even a bit more by hanging onto these arrays as you create them, and avoid the inner-loop [`Lib.nestedProperty` calls below](https://github.com/plotly/plotly.js/pull/1892/files#diff-2c8b7c893b5eec3feb83ac0af6030ccfR194)